internal/dataplane/parser/translators/httproute.go

```diff
@@ -34,100 +43,304 @@ func TranslateHTTPRoute(route *gatewayv1beta1.HTTPRoute) []*HTTPRouteTranslation
+	Rule       gatewayv1beta1.HTTPRouteRule
+	RuleNumber int
+}
+
+// getFiltersKey computes a key from a list of filters.
+// The order of the filters is not important.
+func (m httpRouteRuleMeta) getFiltersKey() string {
+	return getSortedItemsString(m.Rule.Filters)
+}
 
 // getHTTPBackendRefsKey computes a key from a list of backendRefs.
-func getHTTPBackendRefsKey(backendRefs ...gatewayv1beta1.HTTPBackendRef) httpBackendRefsKey {
-	backendKeys := make([]string, 0, len(backendRefs))
+// The order of backedRefs is not important.
+func (m httpRouteRuleMeta) getHTTPBackendRefsKey() string {
+	return getSortedItemsString(m.Rule.BackendRefs)
+}
 
-	// create a list of backend keys
-	for _, backendRef := range backendRefs {
-		var backendKey strings.Builder
+func (m *httpRouteRuleMeta) matches() httpRouteMatchMetaList {
+	matches := make([]httpRouteMatchMeta, 0, len(m.Rule.Matches))
 
-		if backendRef.Group != nil {
-			backendKey.WriteString(string(*backendRef.Group))
-		}
-		backendKey.WriteString(".")
+	for matchNumber, match := range m.Rule.Matches {
+		match := match
+		matches = append(matches, httpRouteMatchMeta{
+			Match:       &match,
+			RuleNumber:  m.RuleNumber,
+			MatchNumber: matchNumber,
+		})
+	}
 
-		if backendRef.Kind != nil {
-			backendKey.WriteString(string(*backendRef.Kind))
+	return matches
+}
+
+type httpRouteMatchMeta struct {
+	Match       *gatewayv1beta1.HTTPRouteMatch
+	RuleNumber  int
+	MatchNumber int
+}
+
+// getKey computes a key from a HTTPRouteMatch, to be used to group rules by their match.
+// The key is derived from the match method, headers and query parameters.
+func (m httpRouteMatchMeta) getKey() string {
+	// According to the spec of HTTPRouteMatch:
+	//
+	// Name is the name of the HTTP Header to be matched. Name matching MUST be
+	// case insensitive. (See https://tools.ietf.org/html/rfc7230#section-3.2).
+	//
+	// If multiple entries specify equivalent header names, only the first
+	// entry with an equivalent name MUST be considered for a match. Subsequent
+	// entries with an equivalent header name MUST be ignored. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered
+	// equivalent.
+	//
+	seenHeaders := make(map[string]struct{})
+	headers := make([]gatewayv1beta1.HTTPHeaderMatch, 0, len(m.Match.Headers))
+	for _, header := range m.Match.Headers {
+		name := strings.ToLower(string(header.Name))
+		header.Name = gatewayv1beta1.HTTPHeaderName(strings.ToLower(string(header.Name)))
+		if _, ok := seenHeaders[name]; ok {
+			continue
 		}
-		backendKey.WriteString(".")
+		seenHeaders[name] = struct{}{}
+		headers = append(headers, header)
+	}
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Name < headers[j].Name
+	})
 
-		if backendRef.Namespace != nil {
-			backendKey.WriteString(string(*backendRef.Namespace))
+	// According to the spec of HTTPQueryParamMatch:
+	//
+	// If multiple entries specify equivalent query param names, only the first
```

Comment:
What does equivalent mean here? Did you mean "the same"?

Reply:
It's direct citation from gateway api docs/comments

Reply:
According to this https://www.rfc-editor.org/rfc/rfc7230#section-3.2, we might need to do normalization